File: Analyzing Baseball Data With R/Book - First Edition/Chapter 7 - Balls and Strikes Effects/ch7_exercise1.py
import pandas as pd
# Run Values of Hits

# In Section 5.8, we found the average run vaalue of a home run and a single

# (a)
# Use similar R code as described in Section 5.8 for the 2011 season data to
# find the mean run values for a double and for a triple

# read in 2011 play by play data
data2011 = pd.read_csv('all2011.csv', header=None)
fields = pd.read_csv('fields.csv')
data2011.columns = fields.loc[:, 'Header']

# We want to calculate the runs scored for the remaining of the inning

# First calculate the runs at a given time
data2011['RUNS'] = data2011.AWAY_SCORE_CT + data2011.HOME_SCORE_CT

# Create a unique ID for each half inning
data2011['HALF_INNING'] = data2011.GAME_ID + data2011.INN_CT.map(str) + \
                            data2011.BAT_HOME_ID.map(str)

# Now we will calculate how many runs were scored at the end of the inning

# Calculate the number of runs scored during each play
data2011['RUNS_SCORED'] = (data2011.BAT_DEST_ID > 3).astype(int) + \
                            (data2011.RUN1_DEST_ID > 3).astype(int) + \
                            (data2011.RUN2_DEST_ID > 3).astype(int) + \
                            (data2011.RUN3_DEST_ID > 3).astype(int)

# Get the Number of runs scored in a specific inning
RUNS_SCORED_INNING = data2011.groupby(['HALF_INNING']).sum().loc[:, 
                                     'RUNS_SCORED']



# Find the total game runs at the beginning of the inning with '[' function 
RUNS_SCORED_START = data2011.loc[:, ['HALF_INNING', 'RUNS']].groupby(
        ['HALF_INNING']).min().loc[:, 'RUNS']

# Get the maximum number of runs in the half inning
MAX = pd.DataFrame(RUNS_SCORED_INNING + RUNS_SCORED_START)
MAX.columns = ['MAX_RUNS']
MAX['half'] = MAX.index

# Merge together the max_runs into the data2011 dataframe
data2011 = data2011.merge(MAX, left_on='HALF_INNING', right_on='half')

# Calculate the runs for the remainder of the inning. Typo in Book
data2011['RUNS_ROI'] = data2011.MAX_RUNS - data2011.RUNS

# Create Binary Variable to determine if runner is on a base before play
RUNNER1 = data2011['BASE1_RUN_ID'].notna().astype(int)
RUNNER2 = data2011['BASE2_RUN_ID'].notna().astype(int)
RUNNER3 = data2011['BASE3_RUN_ID'].notna().astype(int)

# Create the current state 
data2011['STATE'] = RUNNER1.map(str) + RUNNER2.map(str) + RUNNER3.map(str) + \
                    ' ' + data2011.OUTS_CT.map(str)

# Create Binary Variable to determine if a runner is on a base after play
NRUNNER1 = ((data2011.RUN1_DEST_ID == 1) | \
            (data2011.BAT_DEST_ID == 1)).astype(int)

NRUNNER2 = ((data2011.BAT_DEST_ID == 2) | \
            (data2011.RUN1_DEST_ID == 2) | \
            (data2011.RUN2_DEST_ID == 2)).astype(int)

NRUNNER3 = ((data2011.BAT_DEST_ID == 3) | \
            (data2011.RUN1_DEST_ID == 3) | \
            (data2011.RUN2_DEST_ID == 3) | \
            (data2011.RUN3_DEST_ID == 3)).astype(int)

# Get number of outs at the end of play
NOUTS = data2011.OUTS_CT + data2011.EVENT_OUTS_CT

# Get the new state at the end of the play
data2011['NEW_STATE'] = NRUNNER1.map(str) + NRUNNER2.map(str) + \
                        NRUNNER3.map(str) + ' ' + NOUTS.map(str)

# Reduce dataframe to when states change or runs score
data2011 = data2011[(data2011.STATE != data2011.NEW_STATE) | 
            (data2011.RUNS_SCORED > 0)]

# Get the number of outs per inning and merge into the dataframe
OUTS = pd.DataFrame(data2011.loc[:, ['HALF_INNING', 'EVENT_OUTS_CT']].groupby(
        ['HALF_INNING']).sum().loc[:, 'EVENT_OUTS_CT'])
OUTS.columns = ['INNING_OUTS']
OUTS['half'] = OUTS.index
data2011 = data2011.merge(OUTS, left_on='HALF_INNING', right_on='half')

# filter out half innings that are walk-offs because they are not complete
# innings
data2011c = data2011[data2011.INNING_OUTS == 3]

# Calculate the expected number of runs for each element of the matrix
RUNS = pd.DataFrame(data2011c.loc[:, ['STATE', 'RUNS_ROI']].groupby(
        ['STATE']).mean().loc[:, 'RUNS_ROI'])

# Add Three Out States to Create Potential States data frame
s3 = ['000 3', '001 3', '010 3', '011 3', '100 3', '101 3', '110 3', '111 3']
zeros_data = {'RUNS_ROI':[0,0,0,0,0,0,0,0]}
RUNS_POTENTIAL = RUNS.append(pd.DataFrame(data=zeros_data, index=s3))

# Calculate Runs Value of State Before Play
data2011['RUNS_STATE'] = RUNS_POTENTIAL.loc[data2011.
                            STATE,].reset_index().loc[:,'RUNS_ROI']

# Calculate Runs Value of State After Play
data2011['RUNS_NEW_STATE'] = RUNS_POTENTIAL.loc[data2011.
                                NEW_STATE,].reset_index().loc[:,'RUNS_ROI']

# Calculate Runs Value of each play
data2011['RUNS_VALUE'] = data2011.RUNS_NEW_STATE - data2011.RUNS_STATE + \
                            data2011.RUNS_SCORED


# Calculate the value of each count
z = [0,0,0,0,0,0,0,0,0,0,0,0]
counts = ["0-0", "1-0", "2-0", "3-0", "0-1", "1-1", "2-1", "3-1", "0-2", "1-2",
          "2-2", "3-2"]
ct_val = pd.DataFrame({"Value": z})
ct_val.index = counts

d00 = data2011[(data2011.BALLS_CT==0) & (data2011.STRIKES_CT==0)]
d10 = data2011[(data2011.BALLS_CT==1) & (data2011.STRIKES_CT==0)]
d20 = data2011[(data2011.BALLS_CT==2) & (data2011.STRIKES_CT==0)]
d30 = data2011[(data2011.BALLS_CT==3) & (data2011.STRIKES_CT==0)]
d01 = data2011[(data2011.BALLS_CT==0) & (data2011.STRIKES_CT==1)]
d11 = data2011[(data2011.BALLS_CT==1) & (data2011.STRIKES_CT==1)]
d21 = data2011[(data2011.BALLS_CT==2) & (data2011.STRIKES_CT==1)]
d31 = data2011[(data2011.BALLS_CT==3) & (data2011.STRIKES_CT==1)]
d02 = data2011[(data2011.BALLS_CT==0) & (data2011.STRIKES_CT==2)]
d12 = data2011[(data2011.BALLS_CT==1) & (data2011.STRIKES_CT==2)]
d22 = data2011[(data2011.BALLS_CT==2) & (data2011.STRIKES_CT==2)]
d32 = data2011[(data2011.BALLS_CT==3) & (data2011.STRIKES_CT==2)]

# Calculate value for each count
ct_val.loc["0-0", "Value"] = d00.RUNS_VALUE.mean()
ct_val.loc["1-0", "Value"] = d10.RUNS_VALUE.mean()
ct_val.loc["2-0", "Value"] = d20.RUNS_VALUE.mean()
ct_val.loc["3-0", "Value"] = d30.RUNS_VALUE.mean()
ct_val.loc["0-1", "Value"] = d01.RUNS_VALUE.mean()
ct_val.loc["1-1", "Value"] = d11.RUNS_VALUE.mean()
ct_val.loc["2-1", "Value"] = d21.RUNS_VALUE.mean()
ct_val.loc["3-1", "Value"] = d31.RUNS_VALUE.mean()
ct_val.loc["0-2", "Value"] = d02.RUNS_VALUE.mean()
ct_val.loc["1-2", "Value"] = d12.RUNS_VALUE.mean()
ct_val.loc["2-2", "Value"] = d22.RUNS_VALUE.mean()
ct_val.loc["3-2", "Value"] = d32.RUNS_VALUE.mean()

# Calculate the value of balls and strikes in each count

ct_val = pd.DataFrame({"Value": z})
ct_val.index = ["0-0", "1-0", "2-0", "3-0", "0-1", "1-1", "2-1", "3-1", "0-2",
                "1-2", "2-2", "3-2"]

bs_val = pd.DataFrame({"Strike": z, "Ball": z})
bs_val.index = counts

# Mean value of strikeout
d_strikeout = data2011[data2011.EVENT_CD == 3]
strikeout_val = d_strikeout.RUNS_VALUE.mean()

# Mean value of walk
d_walk = data2011[data2011.EVENT_CD == 14]
walk_val = d_walk.RUNS_VALUE.mean()

# Calculate Values of different pitches for Strikes
bs_val.loc["0-0", "Strike"] = round(ct_val.loc["0-1", "Value"] - 
      ct_val.loc["0-0", "Value"], 2)
bs_val.loc["1-0", "Strike"] = round(ct_val.loc["1-1", "Value"] - 
      ct_val.loc["1-0", "Value"], 2)
bs_val.loc["2-0", "Strike"] = round(ct_val.loc["2-1", "Value"] - 
      ct_val.loc["2-0", "Value"], 2)
bs_val.loc["3-0", "Strike"] = round(ct_val.loc["3-1", "Value"] - 
      ct_val.loc["3-0", "Value"], 2)

bs_val.loc["0-1", "Strike"] = round(ct_val.loc["0-2", "Value"] - 
      ct_val.loc["0-1", "Value"], 2)
bs_val.loc["1-1", "Strike"] = round(ct_val.loc["1-2", "Value"] - 
      ct_val.loc["1-1", "Value"], 2)
bs_val.loc["2-1", "Strike"] = round(ct_val.loc["2-2", "Value"] - 
      ct_val.loc["2-1", "Value"], 2)
bs_val.loc["3-1", "Strike"] = round(ct_val.loc["3-2", "Value"] - 
      ct_val.loc["3-1", "Value"], 2)

bs_val.loc["0-2", "Strike"] = round(strikeout_val - 
          ct_val.loc["0-2", "Value"], 2)
bs_val.loc["1-2", "Strike"] = round(strikeout_val - 
          ct_val.loc["1-2", "Value"], 2)
bs_val.loc["2-2", "Strike"] = round(strikeout_val - 
          ct_val.loc["2-2", "Value"], 2)
bs_val.loc["3-2", "Strike"] = round(strikeout_val - 
          ct_val.loc["3-2", "Value"], 2)

# Calculate Values of different pitches for Balls
bs_val.loc["0-0", "Ball"] = round(ct_val.loc["1-0", "Value"] - 
      ct_val.loc["0-0", "Value"], 2)
bs_val.loc["1-0", "Ball"] = round(ct_val.loc["2-0", "Value"] - 
      ct_val.loc["1-0", "Value"], 2)
bs_val.loc["2-0", "Ball"] = round(ct_val.loc["3-0", "Value"] - 
      ct_val.loc["2-0", "Value"], 2)
bs_val.loc["3-0", "Ball"] = round(walk_val - ct_val["3-0", "Value"], 2)

bs_val.loc["0-1", "Ball"] = round(ct_val.loc["1-1", "Value"] - 
      ct_val.loc["0-1", "Value"], 2)
bs_val.loc["1-1", "Ball"] = round(ct_val.loc["2-1", "Value"] - 
      ct_val.loc["1-1", "Value"], 2)
bs_val.loc["2-1", "Ball"] = round(ct_val.loc["3-1", "Value"] - 
      ct_val.loc["2-1", "Value"], 2)
bs_val.loc["3-1", "Ball"] = round(walk_val - ct_val["3-1", "Value"], 2)

bs_val.loc["0-2", "Ball"] = round(ct_val.loc["1-2", "Value"] - 
      ct_val.loc["0-2", "Value"], 2)
bs_val.loc["1-2", "Ball"] = round(ct_val.loc["2-2", "Value"] - 
      ct_val.loc["1-2", "Value"], 2)
bs_val.loc["2-2", "Ball"] = round(ct_val.loc["3-2", "Value"] - 
      ct_val.loc["2-2", "Value"], 2)
bs_val.loc["3-2", "Ball"] = round(walk_val - ct_val.loc["3-2", "Value"], 2)

# (b)
# Compare your values to the ones proposed by John Walsh in the article 
# www.hardballtimes.com/main/article/searching-for-the-games-best-pitch

# The values are similar to those in the fangraphs article. They possibly
# have a smaller magnitude

File: Analyzing Baseball Data With R/Book - First Edition/Chapter 7 - Balls and Strikes Effects/ch7_exercise1.R
rm(list=ls())

# Run Value of Individual Pitches

chapter_path = "C:/Users/conno/Documents/GitHub/Baseball/Analyzing Baseball Data With R/Book - First Edition/Chapter 7 - Balls and Strikes Effects"
setwd(chapter_path)

# (a)
# Calculate the run value of a ball and a strike at any count. For 3-ball and 
# 2-strike conts you need the value of a walk and a strike-out respectively
# (you can calculate them as done for other events in Chapter 5)

# read in 2011 play by play data
data2011 <- read.csv("all2011.csv")
fields <- read.csv("fields.csv")
names(data2011) <- fields[,"Header"]

# We want to calculate the runs scored for the remaining of the inning

# First calculate the runs at a given time
data2011$RUNS <- with(data2011, AWAY_SCORE_CT + HOME_SCORE_CT)

# Create a unique ID for each half inning
data2011$HALF.INNING <- with(data2011, paste(GAME_ID, INN_CT, BAT_HOME_ID))

# Now we will calculate how many runs were scored at the end of the inning

# Calculate the number of runs scored during each play
data2011$RUNS.SCORED <- with(data2011, (BAT_DEST_ID > 3) + (RUN1_DEST_ID > 3)
                             + (RUN2_DEST_ID > 3) + (RUN3_DEST_ID > 3))

# Aggregate to sum the runs scored on each play to determine the number of
# runs scored in the half inning
RUNS.SCORED.INNING <- aggregate(data2011$RUNS.SCORED, 
                                list(HALF.INNING=data2011$HALF.INNING), sum)

# Find the total game runs at the beginning of the inning with "[" function 
RUNS.SCORED.START <- aggregate(data2011$RUNS,
                               list(HALF.INNING=data2011$HALF.INNING), "[", 1)

# Get the maximum number of runs in the half inning
MAX <- data.frame(HALF.INNING=RUNS.SCORED.START$HALF.INNING)
MAX$x <- RUNS.SCORED.INNING$x + RUNS.SCORED.START$x

# Merge the data and name the column
data2011 <- merge(data2011, MAX)
names(data2011)[ncol(data2011)] <- "MAX.RUNS"

# Calculate the runs for the remainder of the inning. Typo in Book
data2011$RUNS.ROI <- with(data2011, MAX.RUNS - RUNS)

# --------------------------------------------------------

# Create Binary Variable to determine if runner is on a base before play
RUNNER1 <- ifelse(as.character(data2011[,"BASE1_RUN_ID"]) == "", 0, 1)
RUNNER2 <- ifelse(as.character(data2011[,"BASE2_RUN_ID"]) == "", 0, 1)
RUNNER3 <- ifelse(as.character(data2011[,"BASE3_RUN_ID"]) == "", 0, 1)

# Create function to get the current state dependent on baserunners and outs
get.state <- function(r1, r2, r3, outs){
  runners <- paste(r1, r2, r3, sep="")
  paste(runners, outs)
}

# Add states to 2011 data
data2011$STATE <- get.state(RUNNER1, RUNNER2, RUNNER3, data2011$OUTS_CT)

# Create Binary Variable to determine if a runner is on a base after play
NRUNNER1 <- with(data2011, as.numeric(RUN1_DEST_ID == 1 |
                                        BAT_DEST_ID == 1))
NRUNNER2 <- with(data2011, as.numeric(RUN1_DEST_ID == 2 |
                                        RUN2_DEST_ID == 2 | BAT_DEST_ID == 2))
NRUNNER3 <- with(data2011, as.numeric(RUN1_DEST_ID == 3 |
                                        RUN2_DEST_ID == 3 | RUN3_DEST_ID == 3 | BAT_DEST_ID == 3))

# Get number of outs at the end of play
NOUTS <- with(data2011, OUTS_CT + EVENT_OUTS_CT)

# Get the new state
data2011$NEW.STATE <- get.state(NRUNNER1, NRUNNER2, NRUNNER3, NOUTS)

# Reduce dataframe to when states change or runs score
data2011 <- subset(data2011, (STATE != NEW.STATE) | (RUNS.SCORED) > 0)

# filter out half innings that are walk-offs because they are not complete
# innings

library(plyr)
data.outs <- ddply(data2011, .(HALF.INNING), summarize, 
                   Outs.Inning=sum(EVENT_OUTS_CT))
data2011 <- merge(data2011, data.outs)
data2011c <- subset(data2011, Outs.Inning==3)

# Calculate the expected number of runs for each element of the matrix
RUNS <- with(data2011c, aggregate(RUNS.ROI, list(STATE), mean))

# Display matrix. First order by out then create matrix

# Get the 5th character to the 5th character in the string
RUNS$OUTS <- substr(RUNS$Group.1, 5, 5)
RUNS <- RUNS[order(RUNS$OUTS),]
RUNS.out <- matrix(round(RUNS$x, 2), 8, 3)
dimnames(RUNS.out)[[2]] <- c("0 outs", "1 out", "2 outs")
dimnames(RUNS.out)[[1]] <- c("000", "001", "010", "011", "100", "101", "110", "111")

# Calculate the run value for each play

# Calculate a potential runs matrix. We know that with 3 outs no more runs can
# be scored
RUNS.POTENTIAL <- matrix(c(RUNS$x, rep(0, 8)), 32, 1)
dimnames(RUNS.POTENTIAL)[[1]] <- c(RUNS$Group.1, "000 3", "001 3", "010 3",
                                   "011 3", "100 3", "101 3", "110 3", "111 3")

# Calculate Runs Value of State Before Play
data2011$RUNS.STATE <- RUNS.POTENTIAL[data2011$STATE,]

# Calculate Runs Value of State After Play
data2011$RUNS.NEW.STATE <- RUNS.POTENTIAL[data2011$NEW.STATE,]

# Calculate Runs Value
data2011$RUNS.VALUE <- data2011$RUNS.NEW.STATE - data2011$RUNS.STATE + 
  data2011$RUNS.SCORED

# Calculate the value of each count
ct.val <- matrix(0, 12, 1)
colnames(ct.val) <- c("Value")
dimnames(ct.val)[[1]] <- c("0-0", "1-0", "2-0", "3-0", "0-1", "1-1", "2-1",
                                "3-1", "0-2", "1-2", "2-2", "3-2")

d00 <- subset(data2011, BALLS_CT==0 & STRIKES_CT==0)
d10 <- subset(data2011, BALLS_CT==1 & STRIKES_CT==0)
d20 <- subset(data2011, BALLS_CT==2 & STRIKES_CT==0)
d30 <- subset(data2011, BALLS_CT==3 & STRIKES_CT==0)
d01 <- subset(data2011, BALLS_CT==0 & STRIKES_CT==1)
d11 <- subset(data2011, BALLS_CT==1 & STRIKES_CT==1)
d21 <- subset(data2011, BALLS_CT==2 & STRIKES_CT==1)
d31 <- subset(data2011, BALLS_CT==3 & STRIKES_CT==1)
d02 <- subset(data2011, BALLS_CT==0 & STRIKES_CT==2)
d12 <- subset(data2011, BALLS_CT==1 & STRIKES_CT==2)
d22 <- subset(data2011, BALLS_CT==2 & STRIKES_CT==2)
d32 <- subset(data2011, BALLS_CT==3 & STRIKES_CT==2)

# Calculate value for each count
ct.val["0-0", "Value"] = mean(d00$RUNS.VALUE)
ct.val["1-0", "Value"] = mean(d10$RUNS.VALUE)
ct.val["2-0", "Value"] = mean(d20$RUNS.VALUE)
ct.val["3-0", "Value"] = mean(d30$RUNS.VALUE)
ct.val["0-1", "Value"] = mean(d01$RUNS.VALUE)
ct.val["1-1", "Value"] = mean(d11$RUNS.VALUE)
ct.val["2-1", "Value"] = mean(d21$RUNS.VALUE)
ct.val["3-1", "Value"] = mean(d31$RUNS.VALUE)
ct.val["0-2", "Value"] = mean(d02$RUNS.VALUE)
ct.val["1-2", "Value"] = mean(d12$RUNS.VALUE)
ct.val["2-2", "Value"] = mean(d22$RUNS.VALUE)
ct.val["3-2", "Value"] = mean(d32$RUNS.VALUE)

# Calculate the value of balls and strikes in each count
bs.val <- matrix(0, 12, 2)
dimnames(bs.val)[[2]] = c("Strike", "Ball")
dimnames(bs.val)[[1]] <- c("0-0", "1-0", "2-0", "3-0", "0-1", "1-1", "2-1", 
                           "3-1", "0-2", "1-2", "2-2", "3-2")

# Mean value of strikeout
d.strikeout <- subset(data2011, EVENT_CD==3)
strikeout_val <- mean(d.strikeout$RUNS.VALUE)

# Mean value of walk
d.walk <- subset(data2011, EVENT_CD==14)
walk_val <- mean(d.walk$RUNS.VALUE)

# Calculate Values of different pitches for Strikes
bs.val["0-0", "Strike"] = round(ct.val["0-1", "Value"] - ct.val["0-0", "Value"], 2)
bs.val["1-0", "Strike"] = round(ct.val["1-1", "Value"] - ct.val["1-0", "Value"], 2)
bs.val["2-0", "Strike"] = round(ct.val["2-1", "Value"] - ct.val["2-0", "Value"], 2)
bs.val["3-0", "Strike"] = round(ct.val["3-1", "Value"] - ct.val["3-0", "Value"], 2)

bs.val["0-1", "Strike"] = round(ct.val["0-2", "Value"] - ct.val["0-1", "Value"], 2)
bs.val["1-1", "Strike"] = round(ct.val["1-2", "Value"] - ct.val["1-1", "Value"], 2)
bs.val["2-1", "Strike"] = round(ct.val["2-2", "Value"] - ct.val["2-1", "Value"], 2)
bs.val["3-1", "Strike"] = round(ct.val["3-2", "Value"] - ct.val["3-1", "Value"], 2)

bs.val["0-2", "Strike"] = round(strikeout_val - ct.val["0-2", "Value"], 2)
bs.val["1-2", "Strike"] = round(strikeout_val - ct.val["1-2", "Value"], 2)
bs.val["2-2", "Strike"] = round(strikeout_val - ct.val["2-2", "Value"], 2)
bs.val["3-2", "Strike"] = round(strikeout_val - ct.val["3-2", "Value"], 2)

# Calculate Values of different pitches for Balls
bs.val["0-0", "Ball"] = round(ct.val["1-0", "Value"] - ct.val["0-0", "Value"], 2)
bs.val["1-0", "Ball"] = round(ct.val["2-0", "Value"] - ct.val["1-0", "Value"], 2)
bs.val["2-0", "Ball"] = round(ct.val["3-0", "Value"] - ct.val["2-0", "Value"], 2)
bs.val["3-0", "Ball"] = round(walk_val - ct.val["3-0", "Value"], 2)

bs.val["0-1", "Ball"] = round(ct.val["1-1", "Value"] - ct.val["0-1", "Value"], 2)
bs.val["1-1", "Ball"] = round(ct.val["2-1", "Value"] - ct.val["1-1", "Value"], 2)
bs.val["2-1", "Ball"] = round(ct.val["3-1", "Value"] - ct.val["2-1", "Value"], 2)
bs.val["3-1", "Ball"] = round(walk_val - ct.val["3-1", "Value"], 2)

bs.val["0-2", "Ball"] = round(ct.val["1-2", "Value"] - ct.val["0-2", "Value"], 2)
bs.val["1-2", "Ball"] = round(ct.val["2-2", "Value"] - ct.val["1-2", "Value"], 2)
bs.val["2-2", "Ball"] = round(ct.val["3-2", "Value"] - ct.val["2-2", "Value"], 2)
bs.val["3-2", "Ball"] = round(walk_val - ct.val["3-2", "Value"], 2)

# (b)
# Compare your values to the ones proposed by John Walsh in the article 
# www.hardballtimes.com/main/article/searching-for-the-games-best-pitch

# The values are similar to those in the fangraphs article. They possibly
# have a smaller magnitude

